### ISSUE TITLE
sunset->previous can return time in future (+patch) [rt.cpan.org #36532]

### ISSUE THREAD

**Issue by jforget** [jforget/DateTime-Event-Sunrise#6]
Migrated from [rt.cpan.org#36532](https://rt.cpan.org/Ticket/Display.html?id=36532) (status was 'resolved')

Requestors:
* michael@cs.mun.ca

Attachments:
* [DateTime-Event-Sunrise.pm-patch.01](https://rt.cpan.org/Ticket/Attachment/471018/232264/DateTime-Event-Sunrise.pm-patch.01)
* [srTest.pl](https://rt.cpan.org/Ticket/Attachment/471018/232267/srTest.pl)

From michael@cs.mun.ca on 2008-06-06 14:17:38
:
```
There is a problem in the calculation of previous/next sunrise/sunset
times: under some conditions, it would return a "previous" sunset time
that was in the future (and also for next sunrises).

  Have attached a test script that shows the problem.
Examples of script:
    $ /tmp/srTest.pl 1212537601 1 1 46.74575,65.2488
    inc=1, count=1, lat=46.74575, long=65.2488
    Error:
            sunSetPrevTime > t
    time = 1212537601 : Tue Jun  3 21:30:01 2008
    sunRisePrevTime = 1212481736 : Tue Jun  3 05:58:56 2008
    sunRiseNextTime = 1212568106 : Wed Jun  4 05:58:26 2008
    sunSetPrevTime  = 1212538107 : Tue Jun  3 21:38:27 2008
    sunSetNextTime  = 1212624556 : Wed Jun  4 21:39:16 2008
and
    $ /tmp/srTest.pl 1212536601 1 1 64.74575,220.2488
    inc=1, count=1, lat=64.74575, long=220.2488
    Error:
            sunRiseNextTime < t
    time = 1212536601 : Tue Jun  3 21:13:21 2008
    sunRisePrevTime = 1212423624 : Mon Jun  2 13:50:24 2008
    sunRiseNextTime = 1212509861 : Tue Jun  3 13:47:41 2008
    sunSetPrevTime  = 1212498208 : Tue Jun  3 10:33:28 2008
    sunSetNextTime  = 1212584787 : Wed Jun  4 10:36:27 2008

For the previous sunset problem: the code computes the sunset
for the "current" day, and if it is greater than the requested time,
returns the sunset time for the previous day.  But if the "current" day
is actually "tomorrow" (which can happen due to the longitude), the
"previous" day's sunset may still be in the future.  A solution is
to add another test and, if needed, compute and return the sunset time for
the "previous previous" day.

The attached patch fixes the two problems shown above and the two other
possible (but not confirmed) similar problems.

Misc info: DateTime::Event::Sunrise, version 0.0501;

perl -v

This is perl, v5.8.5 built for i386-linux-thread-multi

Copyright 1987-2004, Larry Wall

Perl may be copied only under the terms of either the Artistic License
or the
GNU General Public License, which may be found in the Perl 5 source kit.

Complete documentation for Perl, including FAQ lists, should be found on
this system using `man perl' or `perldoc perl'.  If you have access to the
Internet, point your browser at http://www.perl.com/, the Perl Home Page.

os: Linux xxx 2.6.9-67.0.7.ELsmp #1 SMP Sat Mar 15 06:54:55 EDT 2008
i686 i686 i386 GNU/Linux



```

From j2n-forget@orange.fr on 2013-10-18 04:57:55
:
```
Le Ven 06 Juin 2008 16:17:38, mrendell a Ã©critÂ :
> There is a problem in the calculation of previous/next sunrise/sunset
> times: under some conditions, it would return a "previous" sunset time
> that was in the future (and also for next sunrises).

Fixed in the 0.0503 version.
> 
>   Have attached a test script that shows the problem.
> 
Thank you, I used your script to create the t/03nextprev.t
test script.


```

**Comment by jforget** [comment]
Friday 2013-10-18, rt.cpan.org: fixed in 0.0503.

**Comment by jforget** [comment]
Patch and test script provided on Friday 2008-06-06
[DateTime-Event-Sunrise.pm-patch.01.txt](https://github.com/jforget/DateTime-Event-Sunrise/files/5869460/DateTime-Event-Sunrise.pm-patch.01.txt)
[srTest.pl.txt](https://github.com/jforget/DateTime-Event-Sunrise/files/5869461/srTest.pl.txt)
